db/compaction/compaction.cc

```diff
@@ -322,13 +322,65 @@ void Compaction::PopulatePenultimateLevelOutputRange() {
     return;
   }
 
-  int exclude_level =
-      immutable_options_.compaction_style == kCompactionStyleUniversal
-          ? kInvalidLevel
-          : number_levels_ - 1;
+  // exclude the last level, the range of all input levels is the safe range
+  // of keys that can be moved up.
+  int exclude_level = number_levels_ - 1;
+
+  if (immutable_options_.compaction_style == kCompactionStyleUniversal) {
+    exclude_level = kInvalidLevel;
+    std::set<uint64_t> penultimate_inputs;
+    for (const auto& input_lvl : inputs_) {
```

Comment:
Please add a comment explaining what it does. If my understanding is correct, if all files at penultimate level are included in the input, we set exclude_level?

Also, we think we need to document in code comments the whole approach somewhere. it is hard to reason about the whole thing as source code spread widely.